src/interceptors/ClientRequest/utils/normalizeHttpRequestParams.test.ts

```diff
@@ -207,6 +207,7 @@ test('handles [RequestOptions, callback] input', () => {
     method: 'POST',
     protocol: 'https:',
     host: 'mswjs.io',
+    origin: 'https://mswjs.io', // see: https://github.com/mswjs/msw/issues/705
```

Comment:
Strangely, `RequestOptions` doesn't have the `origin` property ([type definitions](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f7ec78508c6797e42f87a4390735bc2c650a1bfd/types/node/http.d.ts#L83-L106)). It once again looks like some arbitrary object is constructed by a third-party library that is not suitable as an input to `ClientRequest`. Unknown properties don't do harm in JavaScript, but they are a culprit when determining what was intended as an input. 

You can verify this yourself:

```ts
// Removed "string" and "URL" as irrelevant inputs.
function test(obj: RequestOptions | LegacyURL): void {}

test({
  origin: '.' // can only specify known properties
})
```

I don't think we should cover each individual (arguably improper) usage of request options. Instead, let's craft a condition for the `URL` case using _other_ properties that URL has (i.e. `hash`). 

```ts
if ('origin' in args[0] && 'hash' in args[0]) {}
```

I think that checking for `instanceof URL` is still a good call, and we should leave it:

```ts
if (args[0] instanceof URL || ('origin' in args[0] && 'hash' in args[0])) {}
```

The bigger this condition grows, the more it asks to be its own function, isolated and unit tested. 

Please, would you be interested in creating this function (i.e. `src/interceptors/ClientRequest/utils/isUrl.ts`)?

Reply:
Where would you see the issues with using `instanceof`? My thinking here is, if you want to guarantee that a `URL` (including all properties and methods like `toString` and `toJSON`) is returned, as defined here:

https://github.com/mswjs/interceptors/blob/16fd35622efe6328497ade15e21a84f6afda0a93/src/interceptors/ClientRequest/utils/normalizeHttpRequestParams.ts#L47

the only check for that really _is_ `instanceof` (or testing for every single property and method), otherwise you won't ever really be sure. Node itself (no Typescript), also only [specifies `Object`](https://nodejs.org/api/http.html#http_http_request_options_callback), so while other options won't be useful, they're also not explicitly forbidde.

IMO, a good way to go would be to use `instanceof` for the `URL` condition, and then tweak the condition for the legacy urls to make sure this really only fits legacy urls. Everything else by my understanding should then be picked up by this case:

https://github.com/mswjs/interceptors/blob/16fd35622efe6328497ade15e21a84f6afda0a93/src/interceptors/ClientRequest/utils/normalizeHttpRequestParams.ts#L123

which then basically does just about what you suggested:

https://github.com/mswjs/interceptors/blob/16fd35622efe6328497ade15e21a84f6afda0a93/src/utils/getUrlByRequestOptions.ts#L91

Reply:
Sounds reasonable. Let's go with the `instanceof` alone and release this is a breaking change. I do hope this eliminates options-related issues in the future. Thanks!

Reply:
Sounds good - thanks for the quick responses!